apps/common/lib/lexical/ast.ex

```diff
@@ -55,6 +55,49 @@ defmodule Lexical.Ast do
     end
   end
 
+  @doc """
+  Traverses the given ast until the given position.
+  """
+  def traverse_until(ast, acc, prewalk_fn, %Position{} = position) do
+    end_line = position.line
+    end_column = position.character
+
+    {_, acc} =
+      ast
+      |> Zipper.zip()
+      |> Zipper.traverse_while(acc, fn zipper, acc ->
+        zipper = maybe_insert_cursor(zipper, position)
+
+        case Zipper.node(zipper) do
+          {_, _, _} = element ->
+            current_line = Sourceror.get_line(element)
+            current_column = Sourceror.get_column(element)
+
+            cond do
+              match?({:__cursor__, _, _}, element) ->
+                {_element, new_acc} = prewalk_fn.(element, acc)
+                {:halt, zipper, new_acc}
+
+              current_line == end_line and current_column > end_column ->
+                {:halt, zipper, acc}
+
+              current_line > end_line ->
+                {:halt, zipper, acc}
+
+              true ->
+                {_element, new_acc} = prewalk_fn.(element, acc)
+                {:cont, zipper, new_acc}
+            end
+
+          element ->
+            {_element, new_acc} = prewalk_fn.(element, acc)
```

Comment:
It's a little odd that the element returned from `prewalk_fn` is ignored in each branch here. I'd say the prewalk function should either only return the new accumulator, or the zipper can be updated in-place with the returned element. 

```elixir
{new_element, new_acc} = prewalk_fn.(element, acc)
new_zipper = Zipper.replace(zipper, new_element)
{:cont, new_zipper, new_acc}
```

Reply:
agreed, you're seeing some evolution from using `Macro.prewalk` to zippers. I hadn't decided what I wanted to do yet.  I'll remove.